pkg/graph/builder.go

```
@@ -632,7 +632,9 @@ func dryRunExpression(env *cel.Env, expression string, resources map[string]*Res
 
 	context := map[string]interface{}{}
 	for resourceName, resource := range resources {
-		context[resourceName] = resource.emulatedObject.Object
+		if resource.emulatedObject != nil {
+			context[resourceName] = resource.emulatedObject.Object
```

should we return an error if the object is nil?

---

Its expected to be nil for the CRD type. So no, because it does not have schema that can be parsed [here](https://github.com/kro-run/kro/blob/4a5b2a310b77fbd43f1babfaac53116d309e8980/pkg/graph/builder.go#L304C35-L304C48).

---

Ohh good catch @Rakhmanov - i wonder why didn't we see these panics when we first tested these RGDs https://github.com/kro-run/kro/blob/main/examples/aws/ack-controller/crds/ecr-public-crd-group.yaml (i have those running in my test cluster w/ kro 0.2.2, no panic observed)

---

@a-hilaly Those are precisely were my starting point and comparison, I could not figure it out why they worked and mine didn't, until ran it with debugger. Since in these CRD's there are 0 usage of cel in any of the resources, they never go down the path of access emulatedObject parameter. And once at least one resource does have usage, you get the panics. I've added the test that should fail on main, and pass on this branch.

---

Nice sleuthing! Indeed, it's the CEL expression usage that triggers this behavior. The graph builder dry-run component assumed that all resources would have an emulated object, which is not the case for CRDs.

---

Now I'm wondering, what would happen if a CEL expression tries to reference a CRD? It can't be validated since CRDs don't have emulated objects. I'm initially thinking about two options
- Allow references to the CRD metadata at least, by injecting a CRD with no versions/schemas.
- Do not allow CEL expressions to reference CRDs (leaning more towards this as a short term option, we can always expand later on)

---

Dang! Nice catch!